### PR TITLE
Mat 161 - add diagonal matrix support through JNI

### PR DESCRIPTION
**Pull request by stuartarchibald**
This PR adds support for diagonal matrix classes.
- Fixes up prototype diagonal matrix C++ code.
- Slightly alters internal storage and `<cinit>` behaviour in java as C/C++ doesn't know about the data lengths, the cost of this change in terms of algorithmic complexity is trivially small as the data sets are essentially vectors. The change essentially zero pads a short diagonal to min(rows,cols) so that loop bounds and algs are cleaner.
- Adds our first end to end test of a java `OGTerminal` `java -> C++ -> java` round trip. This is done as a materialise on `OG*Matrix` and `OG*DiagonalMatrix` classes.

bb pass: [http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/73](http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/73)

### CODE REVIEW

File: src/jshim/jshim.cc [hunk 1]
```diff
@@ -426,6 +476,20 @@ class ExprFactory
         _expr->debug_print();
       }
       break;
+      case OGREALDIAGONALMATRIX_ENUM:
+      {
+        printf("Binding a JOGRealDiagonalMatrix\n");
+        _expr = new JOGRealDiagonalMatrix(&obj);
+        _expr->debug_print();
```
Comment (gmarkall) [inline src/jshim/jshim.cc]: Are the debug_prints still required?

File: src/jshim/jshim.cc [hunk 2]
```diff
@@ -426,6 +476,20 @@ class ExprFactory
...
+      }
+      break;
+      case OGCOMPLEXDIAGONALMATRIX_ENUM:
+      {
+        printf("Binding a JOGComplexDiagonalMatrix\n");
+        _expr = new JOGComplexDiagonalMatrix(&obj);
+        _expr->debug_print();
```
Comment (gmarkall) [inline src/jshim/jshim.cc]: Another debug_print

File: src/jshim/jshim.cc [hunk 3]
```diff
@@ -646,6 +710,7 @@ class DispatchToComplex16ArrayOfArrays: public librdag::Visitor
     };
     void visit(librdag::OGArray<complex16> *thing)
     {
+      printf("Visitor: librdag::OGArray<complex16> branch\n");
```
Comment (gmarkall) [inline src/jshim/jshim.cc:729]: Debugging printf

File: src/jshim/jshim.cc [hunk 4]
```diff
@@ -656,6 +721,7 @@ class DispatchToComplex16ArrayOfArrays: public librdag::Visitor
     }
     void visit(librdag::OGMatrix<complex16> *thing)
     {
+      printf("Visitor: librdag::OGMatrix<complex16> branch\n");
```
Comment (gmarkall) [inline src/jshim/jshim.cc:740]: Debugging printf

File: src/jshim/jshim.cc [hunk 5]
```diff
@@ -426,6 +476,20 @@ class ExprFactory
...
+      case OGREALDIAGONALMATRIX_ENUM:
+      {
+        printf("Binding a JOGRealDiagonalMatrix\n");
+        _expr = new JOGRealDiagonalMatrix(&obj);
+        _expr->debug_print();
+      }
+      break;
+      case OGCOMPLEXDIAGONALMATRIX_ENUM:
+      {
+        printf("Binding a JOGComplexDiagonalMatrix\n");
```
Comment (gmarkall) [inline src/jshim/jshim.cc:494]: debug printf

File: src/jshim/jshim.cc [hunk 6]
```diff
@@ -426,6 +476,20 @@ class ExprFactory
         _expr->debug_print();
       }
       break;
+      case OGREALDIAGONALMATRIX_ENUM:
+      {
+        printf("Binding a JOGRealDiagonalMatrix\n");
```
Comment (gmarkall) [inline src/jshim/jshim.cc:486]: debug printf

File: src/jshim/jshim.cc (hunk 1 above)
Comment (stuartarchibald) [inline src/jshim/jshim.cc]: Nope.
Fix: b66d650
BB Pass. http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/77

File: src/jshim/jshim.cc (hunk 2 above)
Comment (stuartarchibald) [inline src/jshim/jshim.cc]: See b66d650

**Comment by gmarkall** [comment]
This looks generally good - is it a good time to remove or #ifdef out those debugging prints? (I think i noted them all, with the intent of making your life easier if you do want to remove them).

Also, I notice that the OGDiagonalMatrix has a constructor that copies the data, and a public copy constructor that copies the data - I seem to recall that we concluded that copies shoudn't copy data, otherwise we'll make multiple copies when manipulating ASTs - do you agree? (if you do, I'm not suggesting changing it in this PR, but rather that we have a ticket for sorting that out to fix another time).

File: src/jshim/jshim.cc (hunk 6 above)
Comment (stuartarchibald) [inline src/jshim/jshim.cc:486]: These exist in all `case` s. Shall we remove totally or wrap them with #ifdef _DEBUG?

**Comment by stuartarchibald** [comment]
Yes so as noted there are debug `printf`s in all `case` s on that switch. I'm happy to wrap with debug.

**Comment by stuartarchibald** [comment]
489311a wraps debug printf statements.
BB Pass.  http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/78

**Pull request merged**